Step1 - 문자열 덧셈 계산기

wocks1123 opened the pull request:
안녕하세요! 미션수행 후 PR 드립니다.
요구사항에 대한 테스트코드 작성 후 기능 구현하는 식으로 진행했습니다.
기능을 너무 절차적으로 구현한 것 같아 개선하고 싶은데 감이 오지 않습니다…
객체를 활용해야겠다는 생각은 드는데 구체적으로 어디서 어떻게 적용을 해야 할지 잘 모르겠습니다. 어떤 식으로 문제에 접근하고 개선해나가면 좋을지 조언 부탁드립니다.
감사합니다!

wocks1123 committed:
feat: StringCalculator 클래스 생성

wocks1123 committed:
test: StringCalculator 테스트 작성

wocks1123 committed:
test: StringCalculator 구현

- 입력 문자열을 순회하며 구분자를 제외한 숫자 더하는 방식

wocks1123 committed:
feat: 커스텀 구분자 추가

wocks1123 committed:
feat: 입력이 null/빈문자열일 경우 처리 및 검증 추가

wocks1123 committed:
feat: StringCalculator의 `add()`를 static 메서드로 변경

- 상태를 가지지 않기 때문에 static 메서드로 변경 후 StringCalculator 객체생성 방지

wocks1123 committed:
feat: 입력문자열에 숫자 이외의 값 검증

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+            } else if (Character.isDigit(c)) {
+                num = num * 10 + (c - '0');
+            } else {
+                throw new RuntimeException("유효하지 않은 문자 : %s".formatted(c));
```
메세지를 추가해 주신 덕분에 에러원인을 쉽게 파악가능하겠네요!  💯 
추가로 RuntimeException을 직접 사용하기보다 표준 예외를 사용해 보시는건 어떨까요?
- https://velog.io/@chullll/72.-%ED%91%9C%EC%A4%80-%EC%98%88%EC%99%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC

HackSung commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+import org.junit.jupiter.params.provider.NullAndEmptySource;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
```
지난 0단계 요구사항중 아래 내용이 있었는데요
요구사항페이지에 있는 테스트 코드 샘플처럼 AssertJ를 사용하지 않으신 특별한 이유가 있을까요? 🤔 
AssertJ를 사용하시면 가독성이 좋아지거든요.
> JUnit의 Assertions이 아닌 AssertJ의 Assertions를 사용한다.

- https://pjh3749.tistory.com/241

wocks1123 replied:
별 고민없이 똑같지 않나?하고 구분없이 사용했는데 AssertJ가 좀 더 가독성이 좋고 인텔리센스 활용하기 좋은 것 같습니다!

HackSung commented:
> 안녕하세요! 미션수행 후 PR 드립니다. 요구사항에 대한 테스트코드 작성 후 기능 구현하는 식으로 진행했습니다. 

네~ TDD방식으로 개발하셨군요? 
좋은 개발 습관을 가지신 것 같아요!  👍

> 기능을 너무 절차적으로 구현한 것 같아 개선하고 싶은데 감이 오지 않습니다… 
객체를 활용해야겠다는 생각은 드는데 구체적으로 어디서 어떻게 적용을 해야 할지 잘 모르겠습니다. 

저도 오랫동안 C언어를 사용해 절차적 방식 프로그래밍만 하다가 Java로 전향하고 나서 한동안 막막했던 기억이 있어 충분히 공감이 갑니다.
시간이 지나면서 점점 객체지향프로그래밍에 익숙해 졌고 재사용성, 확장성, 유지보수 및 테스트 용이성 등의 객체지향설계의 효능을 직접 느끼고 나니
객체지향 프로그래밍에 더 매력을 느끼게 되었고 지금은 이러한 설계 철학이 많은 개발 과제에서 얼마나 큰 가치를 발휘하는지 실감하고 있습니다. 
처음에는 낯설고 어렵게 느껴졌던 개념이었지만 꾸준히 학습하고 경험을 쌓으니 자연스럽게 익숙해지는 것 같습니다. 😄 

> 어떤 식으로 문제에 접근하고 개선해나가면 좋을지 조언 부탁드립니다. 감사합니다!

정답이 있는건 아니지만 저는 대략 다음과 같은 순서로 접근하고 있습니다.
1. 요구사항 분석을 통해 필요한 기능들을 도출한다.
2. 도출된 기능을 책임질 클래스(속성, 메서드)들을 설계한다. (가능한 단일책임원칙(SRP)을 지킨다.)
3. 클래스간 관계(상속, 구성)와 통신항목을 정의한다.
4. 중복된 코드를 제거하거나 변경 가능성이 높은 부분은 인터페이스로 분리하는 등 리팩토링을 수행한다.

이번 문자열 덧셈 계산기 미션도 필요한 기능은 알고계시니 책임(기능)기반으로 클래스를 분리하는 것부터 시작 해보시면 좋을 것 같습니다.
처음부터 완벽할 수는 없으니 제가 옆에서 최대한 도와드리겠습니다.

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+    private StringCalculator() {
+    }
+
+    public static int add(String text) {
```
메서드명은 `add()`지만 하나의 메서드 안에 모든 기능들이 들어가 있네요?
물론 기능들이 간단해 슈퍼 메서드 하나로 끝내는게 나아 보일수 있지만
객체지향설계를 훈련해 본다는 마음으로 
입력 유효성체크, 파싱, 합산 등 기능별로 작은 클래스나 메서드로 적절히 나눠보는 건 어떨까요?

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+
+        int sum = 0;
+        int num = 0;
+        String delimiters = ",:";
```
숫자나 문자열과 같은 하드코딩된 값을 상수로 대체하면 코드의 가독성을 높이고 의미를 명확히 할 수 있을 것 같아요.

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+            if (delimiters.indexOf(c) >= 0) {
+                sum += num;
+                num = 0;
+            } else if (Character.isDigit(c)) {
```
다음 규칙을 지키려면 어떻게 해야 할까요? 🤔 
> 객체 지향 생활 체조 원칙 2: else 예약어를 쓰지 않는다

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+
+        if (text.startsWith("//") && "\\n".equals(text.substring(3, 5))) {
+            delimiters += text.charAt(2);
+            text = text.substring(5);
```
필수는 아니지만 Java의 정규식 패턴 매칭을 통한 추출방법도 고려해 보세요

HackSung reviewed (changes requested):
미션수행하시느라 수고 많으셨습니다. 👏 
피드백 남겨 드렸으니 검토부탁드려요 🙏

wocks1123 committed:
refactor: `assertEquals()`를 `assertThat()`으로 수정

wocks1123 committed:
feat: 구분자 추출을 수행하는 클래스 구현

wocks1123 committed:
feat: 문자열에서 숫자를 추출하는 클래스 구현

wocks1123 committed:
refactor: StringCalculator 클래스 수정

- 역할을 분리
  - DelimiterGroup : 입력 문자열에서 커스텀 구분자를 분리
  - NumberExtractor
      - 커스텀 구분자를 분리한 문자열에서 숫자를 추출
      - 숫자가
  - StringCalculator : 각 객체를 사용해 합 계산
- 테스트 코드 수정

wocks1123 committed:
refactor: NumberExtractor의 DelimiterGroup 의존성 제거

-  DelimiterGroup의 역할을 분명하게 수정
  - 커스텀 구분자 목록을 관리한다.
  - 입력 문자열에서 커스텀 구분자를 제거한 문자열을 가진다.

wocks1123 committed:
refactor: NumberExtractor, NumberValidator 분리

- NumberValidator : 입력 문자열의 숫자를 검증
- NumberExtractor를 인터페이스로 변경 : 문자열 추출 방법 확장성 고려

wocks1123 committed:
refactor: 클래스 이름 변경 DelimiterGroup -> InputText

- 역할에 맞는 명확한 이름으로 변경

wocks1123 committed:
refactor: InputText와 CustomDelimiterExtractor 분리

- InputText는 숫자 문자열과 구분자를 가진 레코드로 변경
- 커스텀 구분자를 출하는 CustomDelimiterExtractor 클래스 추가

wocks1123 committed:
refactor: NumberValidator를 인터페이스로 분리

wocks1123 committed:
refactor: CustomDelimiterExtractor를 인터페이스로 분리

wocks1123 committed:
refactor: NumberExtractor 이름을 TextSplitter로 변경

- 테스트 추가

wocks1123 committed:
feat: 제약추가 - InputText는 null값을 가질 수 없다.

- 관련 테스트항목 추가

wocks1123 commented:
피드백 감사합니다! 리뷰해주신 부분 반영해서 수정했습니다.
최대한 객체를 활용하도록 코드를 수정했습니다. 기능을 큰 부분으로 나누어서 클래스로 추출하니 다시 분리가 가능한 부분이 보이더라고요. 하나씩 분리해나가며 각 클래스가 자기 역할을 수행하도록 구현했습니다. 책임을 분리하니 어느 부분을 테스트할지 명확히 보여서 좋았던 것 같습니다. 확장성을 고려해서 여러 부분을 인터페이스로 분리했는데 다시보니 `Validator` 까지도 확장이 필요한가 생각도 드네요.
진행하면서 역할을 어떻게 나누는게 적절할지 고민하는 과정이 힘들었네요. 이번엔 간단한 과제였지만 규모가 더 상황에서도 이런 방식으로 접근해나가면 될까요?

HackSung commented on src/main/java/calculator/SimpleNumberValidator.java:
```diff
@@ -0,0 +1,20 @@
+
+public class SimpleNumberValidator implements NumberValidator {
+
+    private static String DIGIT_REGEX_PATTE = "\\d+";
```
상수의 불변성 보장을 위한 `final` 키워드가 누락되었네요? 
상수명도 끝이 마무리가 안된 것처럼 보여요. 🤔

HackSung commented on src/main/java/calculator/SimpleNumberValidator.java:
```diff
@@ -0,0 +1,20 @@
+    private static String DIGIT_REGEX_PATTE = "\\d+";
+
+
+    public void validateNumbers(String[] splittedString) {
```
`@Override` 명시가 빠진 것 같네요?

HackSung commented on src/main/java/calculator/SimpleTextSplitter.java:
```diff
@@ -0,0 +1,17 @@
+            return new String[]{inputText.numberText()};
+        }
+
+        Pattern pattern = Pattern.compile("[%s]".formatted(inputText.delimiters()));
```
Pattern 객체는 생성비용이 높은 것으로 알려져 있는데 
입력받을때 마다 매번 생성하지 않을 수 있는 방법이 있을까요?
아래 글을 참고하여 개선해 볼 수 있을까요?
- https://zhenxikim.tistory.com/79

wocks1123 replied:
`Map<String, Pattern> patternCache`를 만들어서 한번 컴파일된 패턴을 재사용하도록 수정했습니다!

HackSung commented on src/test/java/calculator/SimpleCustomDelimiterExtractorTest.java:
```diff
@@ -0,0 +1,72 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class SimpleCustomDelimiterExtractorTest {
```
와우! 구현하기 귀찮을 수도 있는데 클래스별로 테스트케이스를 정말 꼼꼼하게 작성하셨네요! 💯

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,38 @@
+
+        InputText inputText = customDelimiterExtractor.extractDelimiter(text);
+
+        String[] numbers = textSplitter.splitText(inputText);
```
이번 미션을 여기서 멈춰도 괜찮지만 더 높은 완성도 도전을 위한 추가제안을 드려 볼게요.
현재 입력에서 추출한 값을 String 타입으로 처리하고 있는데
양수를 보장하는 새로운 타입의 클래스(VO)를 하나 설계하여 적용해 보시고 어떤 점이 좋아지는지 공유부탁드릴께요.
관련하여 향후 미션에서도 자주 사용하게 될 VO와 일급컬렉션에 대한 아래 글을 미리 공부하면 도움이 되실 것 같아요.
혹시나 바쁘셔서 시간이 부족하다면 추가제안 구현하지 않고 이번 미션을 마무리 하셔도 괜찮습니다. 😄 

- 값객체: https://medium.com/29cm/%EA%B0%92-%EA%B0%9D%EC%B2%B4-value-object-%EB%A5%BC-%ED%99%9C%EC%9A%A9%ED%95%98%EC%97%AC-%EB%B3%80%EA%B2%BD-%EC%9A%A9%EC%9D%B4%EC%84%B1-%EA%B0%9C%EC%84%A0%ED%95%98%EA%B8%B0-7457787b84c0
              
- 일급컬렉션: https://jojoldu.tistory.com/412

wocks1123 replied:
기존에는 구분자에서 추출한 숫자들을 `String[]`으로 처리했는데 이 부분을 수를 나타내는 `PositiveNumber`와 숫자들을 모은 `Numbers` 객체로 풀었습니다. 또 너무 많은 `PositiveNumber` 객체를 생성하지 않도록 캐시를 추가했습니다.
이렇게 VO를 사용하도록 코드를 변경하니 이 객체를 사용하는 곳에서는 이미 적절한 값을 가지고 있다는걸 확신할 수 있는 점이 좋았습니다. 객체 생성 과정에서 검증을 하고 이후에는 값을 수정할 수 없다는 특성이 메서드 호출간 객체를 주고받는 과정에 안정성을 주는 것 같습니다. Numbers가 유효한걸 보장하니 `StringCalculator.add()` 메서드에 있던 validator가 필요 없어져 제거할 수 있었습니다.

HackSung commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,38 @@
+            return 0;
+        }
+
+        InputText inputText = customDelimiterExtractor.extractDelimiter(text);
```
extractDelimiter라는 메서드명만 보면 반환명이 delimiter가 아닌 inputText이다 보니
코드를 처음 보는 사람이 약간 혼동할 수도 있을 것 같은데 어떻게 생각하세요? 🤔

wocks1123 replied:
`InputText` 클래스를 추가하면서 메서드 반환 타입만 바꿨는데 이걸 사용하는 메서드 이름이 어색했네요.  `DelimiterExtractor`를 `DelimiterParser`로 변경했습니다.

HackSung commented:
> 피드백 감사합니다! 리뷰해주신 부분 반영해서 수정했습니다. 최대한 객체를 활용하도록 코드를 수정했습니다. 기능을 큰 부분으로 나누어서 클래스로 추출하니 다시 분리가 가능한 부분이 보이더라고요. 하나씩 분리해나가며 각 클래스가 자기 역할을 수행하도록 구현했습니다. 책임을 분리하니 어느 부분을 테스트할지 명확히 보여서 좋았던 것 같습니다. 확장성을 고려해서 여러 부분을 인터페이스로 분리했는데 다시보니 `Validator` 까지도 확장이 필요한가 생각도 드네요. 진행하면서 역할을 어떻게 나누는게 적절할지 고민하는 과정이 힘들었네요. 이번엔 간단한 과제였지만 규모가 더 상황에서도 이런 방식으로 접근해나가면 될까요?

일단 이번 미션에서 객체 지향 설계의 이점을 체감하신 게 가장 큰 수확인 것 같습니다.
다만 말씀하신 것처럼 모든 책임을 분리하고 확장 가능하게 하다 보면
대체 어디까지 분리해야 하는지 더 복잡해지는 건 아닌지 하는 고민에 빠지게 되는 것 같아요.
저의 경우 현재 요구사항만 보았을 때 확장 가능성이 없다고 판단되면 인터페이스 없이 간단히 구현체로만 개발하고 
필요한 시점에 리팩토링으로 인터페이스를 도입합니다.
규모가 큰 프로젝트에서도 기본적으로 지금과 같은 방식으로 접근해 나가는 것이 좋다고 생각하지만
프젝트의 상황과 복잡성에 따라 모듈과 패키지의 구조에 대한 아키텍처(레이어드, 클린 아키텍처 등)까지 고민이 필요 하겠지요.
향후 수업과 미션으로 배우시게 될 도메인 주도 설계가 바로 대규모 프로젝트의 복잡성을 다스리기 위한 방법중의 하나입니다. 😃

HackSung reviewed (changes requested):
와우! 피드백을 훌륭하게 반영하신 것 같아요! 👍 
좀 더 심화된 코멘트를 남겨드렸으니 이것도 검토 부탁드려요~

wocks1123 committed:
fix: 오타수정, final 누락 수정

wocks1123 committed:
feat: 숫자를 표현하는 객체 추가

- 기존 String[]으로 표현하던 계산 숫자를 객체로 변경
  - PositiveNumber : 계산기 요구사항에 맞게 음수를 가질 수 없음
  - StringNumbers : 숫자 배열을 관리
- 관련 테스트 코드 수정

wocks1123 committed:
remove: 불필요 코드 삭제

wocks1123 committed:
rename: DelimiterExtractor -> DelimiterParser

wocks1123 committed:
feat: PositiveNumber에 캐시 적용

wocks1123 committed:
rename: StringNumbers -> Numbers

wocks1123 committed:
feat: PatternCache 추가

- `Pattern.compile()` 메서드 호출을 최소화하기 위한 캐시 추가

wocks1123 commented:
피드백 반영해서 수정했습니다. 미처 확인 못한 부분도 있었고 새롭게 알게된 부분도 많았네요. VO 객체에 대한 개념은 어렴풋이 알고 다른 사람의 코드를 볼 때도 인지는 했지만 막상 제가 코드 작성하면서 떠올리지는 못했네요. 이번에 VO 클래스를 적용해보면서 필요성에 대해 공부가 많이 된 것 같습니다.

HackSung reviewed (approved):
의도한대로 피드백 반영을 아주 잘해주셨습니다! 💯 
이제 다음 단계에서 뵙도록 하지요. 🚀
